tests/lib/modem_trace/modem_trace/src/modem_trace_test.c

```diff
@@ -0,0 +1,312 @@
+
+	TEST_ASSERT_EQUAL(0, nrf_modem_lib_trace_init(&trace_heap));
+}
+
+void test_modem_trace_start_coredump_only(void)
+{
+	nrf_modem_at_printf_ExpectTraceModeAndReturn(NRF_MODEM_LIB_TRACE_COREDUMP_ONLY, 0);
+	TEST_ASSERT_EQUAL(0, nrf_modem_lib_trace_start(NRF_MODEM_LIB_TRACE_COREDUMP_ONLY));
+}
+
+void test_modem_trace_start_all(void)
+{
+	nrf_modem_at_printf_ExpectTraceModeAndReturn(NRF_MODEM_LIB_TRACE_ALL, 0);
+	TEST_ASSERT_EQUAL(0, nrf_modem_lib_trace_start(NRF_MODEM_LIB_TRACE_ALL));
+}
+
+void test_modem_trace_start_ip_only(void)
+{
+	nrf_modem_at_printf_ExpectTraceModeAndReturn(NRF_MODEM_LIB_TRACE_IP_ONLY, 0);
+	TEST_ASSERT_EQUAL(0, nrf_modem_lib_trace_start(NRF_MODEM_LIB_TRACE_IP_ONLY));
+}
+
+void test_modem_trace_start_lte_ip(void)
+{
+	nrf_modem_at_printf_ExpectTraceModeAndReturn(NRF_MODEM_LIB_TRACE_LTE_IP, 0);
+	TEST_ASSERT_EQUAL(0, nrf_modem_lib_trace_start(NRF_MODEM_LIB_TRACE_LTE_IP));
+}
+
+/* Test that traces are forwarded to the trace medium when the module is correctly initialized
+ * and nrf_modem_lib_trace_start has been called.
+ */
+void test_modem_trace_write_to_medium(void)
+{
+	int ret;
+	const uint8_t sample_trace_data[10];
+
+	__wrap_trace_medium_init_ExpectAndReturn(0);
+	TEST_ASSERT_EQUAL(0, nrf_modem_lib_trace_init(&trace_heap));
+
+	nrf_modem_at_printf_ExpectTraceModeAndReturn(NRF_MODEM_LIB_TRACE_ALL, 0);
+	TEST_ASSERT_EQUAL(0, nrf_modem_lib_trace_start(NRF_MODEM_LIB_TRACE_ALL));
+
+	/* Simulate reception of a modem trace. */
+	ret = nrf_modem_lib_trace_process(sample_trace_data, sizeof(sample_trace_data));
+	TEST_ASSERT_EQUAL(0, ret);
+
+	__wrap_trace_medium_write_ExpectAndReturn(sample_trace_data, sizeof(sample_trace_data),
+					   sizeof(sample_trace_data));
+	trace_processed_callback_ExpectAndReturn(sample_trace_data, sizeof(sample_trace_data), 0);
+
+	/* Let trace thread run. */
+	k_sleep(K_MSEC(1));
+}
+
+/* With the module correctly initialized, test that traces are forwarded to the trace medium
+ * even when nrf_modem_lib_trace_start has not yet been called.
+ */
+void test_modem_trace_process_when_modem_trace_start_was_not_called(void)
+{
+	int ret;
+	const uint8_t sample_trace_data[10];
+
+	__wrap_trace_medium_init_ExpectAndReturn(0);
+	TEST_ASSERT_EQUAL(0, nrf_modem_lib_trace_init(&trace_heap));
+
+	/* Simulate reception of a modem trace. */
+	ret = nrf_modem_lib_trace_process(sample_trace_data, sizeof(sample_trace_data));
+	TEST_ASSERT_EQUAL(0, ret);
+
+	__wrap_trace_medium_write_ExpectAndReturn(sample_trace_data, sizeof(sample_trace_data),
+					   sizeof(sample_trace_data));
+	trace_processed_callback_ExpectAndReturn(sample_trace_data, sizeof(sample_trace_data), 0);
+
+	/* Let trace thread run. */
+	k_sleep(K_MSEC(1));
+}
+
+/* Test that trace_medium_write is not called when trace module fails to initialize, and that
+ * nrf_modem_trace_processed_callback is called even when initialization fails.
+ */
+void test_modem_trace_write_when_medium_init_fails(void)
```

Comment:
Could this be split into one test for `nrf_modem_lib_trace_init_ebusy`, one for `nrf_modem_lib_trace_start_enxio` and one for `nrf_modem_lib_trace_process_enxio` (which is basically the same as the next test)? 

You could call the first one from the second if that is required for nrf_modem_lib_trace_start to return -ENXIO etc.